core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -403,6 +406,13 @@ private void checkNestedFieldsLimit(Map<String, ObjectMapper> fullPathObjectMapp
         }
     }
 
+    private void checkTotalFieldsLimit(long totalMappers) {
+        long allowedTotalFields = indexSettings.getValue(INDEX_MAPPING_TOTAL_FIELDS_LIMIT_SETTING);
+        if (allowedTotalFields > 0 && allowedTotalFields < totalMappers) {
```

Comment:
Like Simon suggested, I would just do: `if (allowedTotalFields < totalMappers) {`. If someone needs to have many fields anyway and understands the issue, (s)he can still set `index.mapping.total_fields.limit=1000000` for instance.